include/nbl/builtin/hlsl/bxdf/base/cook_torrance_base.hlsl

```diff
@@ -0,0 +1,378 @@
+        const vector3_type localV = interaction.getTangentSpaceV();
+        const vector3_type localH = ndf.generateH(localV, u);
+        const scalar_type VdotH = hlsl::dot(localV, localH);
+        const vector3_type H = hlsl::mul(interaction.getFromTangentSpace(), localH);
+
+        ray_dir_info_type L;
+        if (scalar_type(2.0) * VdotH * localH.z > localV.z) // NdotL>0, compiler's Common Subexpression Elimination pass should re-use 2*VdotH later
+        {
+            assert(VdotH >= scalar_type(0.0));
+            ray_dir_info_type V = interaction.getV();
+            struct reflect_wrapper  // so we don't recalculate VdotH
+            {
+                vector3_type operator()() NBL_CONST_MEMBER_FUNC
+                {
+                    return r(VdotH);
+                }
+                bxdf::Reflect<scalar_type> r;
+                scalar_type VdotH;
+            };
+            reflect_wrapper rw;
+            rw.r = bxdf::Reflect<scalar_type>::create(V.getDirection(), H);
+            rw.VdotH = VdotH;
+            L = V.template reflect<reflect_wrapper>(rw);
+
+            cache = anisocache_type::createForReflection(localV, localH);
+        }
+        else    // fail if samples have invalid paths
+            L.makeInvalid();    // should check if sample direction is invalid
+
+        const vector3_type T = interaction.getT();
+        const vector3_type B = interaction.getB();
+        const vector3_type _N = interaction.getN();
+
+        return sample_type::create(L, T, B, _N);
+    }
+    template<typename C=bool_constant<IsBSDF> >
+    enable_if_t<C::value && IsBSDF, sample_type> generate(NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, const vector3_type u, NBL_REF_ARG(anisocache_type) cache)
+    {
+        const vector3_type localV = interaction.getTangentSpaceV();
+        const scalar_type NdotV = localV.z;
+
+        fresnel_type _f = impl::getOrientedFresnel<fresnel_type, IsBSDF>::__call(fresnel, NdotV);
+        fresnel::OrientedEtaRcps<monochrome_type> rcpEta = _f.getOrientedEtaRcps();
+
+        const vector3_type upperHemisphereV = ieee754::flipSignIfRHSNegative<vector3_type>(localV, hlsl::promote<vector3_type>(NdotV));
+        const vector3_type localH = ndf.generateH(upperHemisphereV, u.xy);
+        const scalar_type VdotH = hlsl::dot(localV, localH);
+        const vector3_type H = hlsl::mul(interaction.getFromTangentSpace(), localH);
+
+        assert(NdotV * VdotH > scalar_type(0.0));
+        const scalar_type reflectance = _f(hlsl::abs(VdotH))[0];
+
+        scalar_type rcpChoiceProb;
+        scalar_type z = u.z;
+        bool transmitted = math::partitionRandVariable(reflectance, z, rcpChoiceProb);
+
+        ray_dir_info_type V = interaction.getV();
+        const vector3_type _N = interaction.getN();
+        Refract<scalar_type> r = Refract<scalar_type>::create(V.getDirection(), H);
+        const scalar_type LdotH = hlsl::mix(VdotH, r.getNdotT(rcpEta.value2[0]), transmitted);
```

Comment:
`r.getNdotT(rcpEta.value2[0])` not optimal
https://github.com/Devsh-Graphics-Programming/Nabla/blob/069f49960677178aab91492d4b731ea761273134/include/nbl/builtin/hlsl/bxdf/fresnel.hlsl#L227-L233